source/adapters/level_zero/command_buffer.cpp

```diff
@@ -706,8 +793,78 @@ urCommandBufferReleaseExp(ur_exp_command_buffer_handle_t CommandBuffer) {
   return UR_RESULT_SUCCESS;
 }
 
-UR_APIEXPORT ur_result_t UR_APICALL
-urCommandBufferFinalizeExp(ur_exp_command_buffer_handle_t CommandBuffer) {
+namespace {
+ur_result_t
+commandBufferFinalizeNew(ur_exp_command_buffer_handle_t CommandBuffer) {
+
+  UR_ASSERT(CommandBuffer, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  // It is not allowed to append to command list from multiple threads.
+  std::scoped_lock<ur_shared_mutex> Guard(CommandBuffer->Mutex);
+
+  // TODO Maybe the ResetCommandList could be run at the end to pre-emptively
+  // prepare the events for the next execution of the graph. In that case, we
+  // need to add a barrier to wait on ExecutionFinishedEvent at the start of the
+  // ResetCommandList.
+  if (!CommandBuffer->IsInOrderCmdList) {
+    // Reset the L0 events we use for command-buffer sync-points to the
+    // non-signaled state. This is required for multiple submissions.
+    for (auto &Event : CommandBuffer->ZeEventsList) {
+      ZE2UR_CALL(zeCommandListAppendEventReset,
+                 (CommandBuffer->ZeCommandListResetEvents, Event));
+    }
+  }
+
+  // Wait for the Copy Queue to finish at the end of the compute command list.
+  if (CommandBuffer->useCopyEngine()) {
+    ZE2UR_CALL(zeCommandListAppendEventReset,
+               (CommandBuffer->ZeCommandListResetEvents,
+                CommandBuffer->CopyFinishedEvent->ZeEvent));
+
+    ZE2UR_CALL(zeCommandListAppendBarrier,
+               (CommandBuffer->ZeComputeCommandList, nullptr, 1,
+                &CommandBuffer->CopyFinishedEvent->ZeEvent));
+  }
+
+  // Reset the all-reset-event for the UR command-buffer that is signaled
+  // when all events of the main command-list have been reset.
+  ZE2UR_CALL(zeCommandListAppendEventReset,
+             (CommandBuffer->ZeComputeCommandList,
+              CommandBuffer->AllResetEvent->ZeEvent));
+
+  // At the moment, only the profiling command-list has a wait on
+  // ExecutionFinishedEvent
+  if (CommandBuffer->IsProfilingEnabled) {
+    ZE2UR_CALL(zeCommandListAppendEventReset,
+               (CommandBuffer->ZeCommandListResetEvents,
+                CommandBuffer->ExecutionFinishedEvent->ZeEvent));
+
+    ZE2UR_CALL(zeCommandListAppendSignalEvent,
+               (CommandBuffer->ZeComputeCommandList,
+                CommandBuffer->ExecutionFinishedEvent->ZeEvent));
+  }
+
+  ZE2UR_CALL(zeCommandListAppendSignalEvent,
+             (CommandBuffer->ZeCommandListResetEvents,
+              CommandBuffer->AllResetEvent->ZeEvent));
+
+  // Close the command lists and have them ready for dispatch.
+  ZE2UR_CALL(zeCommandListClose, (CommandBuffer->ZeComputeCommandList));
+  ZE2UR_CALL(zeCommandListClose, (CommandBuffer->ZeCommandListResetEvents));
+
+  if (CommandBuffer->useCopyEngine()) {
+    ZE2UR_CALL(zeCommandListAppendSignalEvent,
```

Comment:
I think this might need to be a barrier that waits on all the commands in the copy command-list before signaling.